staging/src/k8s.io/apiserver/pkg/endpoints/filters/read_write_deadline_test.go

```diff
@@ -0,0 +1,1134 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package filters
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+	responsewritertesting "k8s.io/apiserver/pkg/endpoints/responsewriter/testing"
+)
+
+func TestPerRequestWriteDeadlineWithWriteShouldReturnErrorAfterTimeout(t *testing.T) {
+	// This test documents the behavior of the per request write deadline
+	// using a standard net http server.
+	//
+	// scenario:
+	//  - write deadline is set to 100ms
+	//  - the handler waits until the write deadline exceeds
+	//  - after timeout occurs, the handler keeps writing 1kB of data at a time
+	// we expect the Write method of the ResponseWriter object to return an
+	// i/o timeout error once its internal buffer is full.
+	t.Parallel()
+
+	tests := []struct {
+		protoMajor        int
+		waiter            waiter
+		verifyClientErrFn func(*testing.T, error)
+	}{
+		{
+			protoMajor: 1, // http/1x
+			// write timeout is set to 100ms, see below; a wait of
+			// 5s should be enough to withstand flakes in CI.
+			waiter: &waitWithDuration{after: 5 * time.Second},
+			verifyClientErrFn: func(t *testing.T, err error) {
+				t.Helper()
+				// TODO: look into why we see this error
+				shouldContain := "local error: tls: bad record MAC"
+				if !strings.Contains(err.Error(), shouldContain) {
+					t.Errorf("expected the error to contain: %q, but got: %v", shouldContain, err)
+				}
+			},
+		},
+		{
+			protoMajor: 2, // http/2.0
+			waiter:     &waitWithChannelClose{after: make(chan time.Time)},
+			verifyClientErrFn: func(t *testing.T, err error) {
+				t.Helper()
+				if !responsewritertesting.IsStreamReadOrWriteTimeout(err) {
+					t.Errorf("expected a stream reset error, but got: %v", err)
+				}
+			},
+		},
+	}
+	for _, test := range tests {
+		t.Run(fmt.Sprintf("HTTP/%d", test.protoMajor), func(t *testing.T) {
+			handlerDoneCh := make(chan error, 1)
+			server := httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+				defer close(handlerDoneCh)
+				if req.ProtoMajor != test.protoMajor {
+					t.Errorf("expected ProtoMajor: %d, but got: %s", test.protoMajor, req.Proto)
```

Comment:
this is another go routine, isn't it? I believe you are not allowed to call `t.Errorf`.

Reply:
I thought calling `t.Errorf` is safe as long as it is called before the test completes, but i am seeing a data race when the test uses `t.Parallel()` and has sub tests.  I will investigate the data race issue I am seeing, for now i am calling `t.Errorf` from within the test.